### DISCUSSION TITLE
Help with writing parentheses rules

### DISCUSSION THREAD

**Discussion by katawful** [tree-sitter/tree-sitter#1132] [Q&A] [answered]
I've been working on implementing parentheses rules for a language that uses them solely for precedence (i.e. they're not needed for any statement/expression). [This](https://github.com/katawful/tree-sitter-obse/tree/refactor) is the repo/branch with the parser I'm creating

My current problem is getting the parentheses to be recognized in expressions with multiple operands/operators like so:
```
if (x > y || z) < 1
if (x > y) || (y < z)
if (x > y || y < z)
```
I can only seem to get the last or the first working and no other combination working. I'm not sure how to properly approach this kind of parentheses

**Comment by ahelwer** [comment, marked as answer]
Usually I have a central "expression" rule which is mutually-recursive with a bunch of other rules, like:

```
_expr: $ => choice(
  $.identifier,
  $.gt,
  $.lt,
  $.lor
  $.parentheses
),

identifier: $ => /\w+/,

gt: $ => seq($._expr, '>', $._expr),

lt: $ => seq($._expr, '<', $._expr),

lor: $ => seq($._expr, '||', $._expr),

parentheses: $ => seq('(', $._expr, ')')
```

Then you define precedence between infix operators however you like. How are you trying to do it?

**Reply by katawful** [reply]
I only nested expressions as a whole, something like:

```js
_expressions: $ => choice(
  $.binary_expr,
  $.paren_binary_expr,
),

binary_expr: $ => seq(
  $.operand,
  $.operator,
  $.operand,
),

paren_binary_expr: $ => seq(
  '(',
  $.binary_expr,
  ')',
),
```
This way got me the closest to a solution, but I didn't try nesting one level up like your example. I will try that

**Reply by katawful** [reply]
Ok yeah that method works, but it doesn't work when I need to string together multiple operations like so:
```
let x = (y + z - a) * b
```

**Reply by ahelwer** [reply]
Does it work when you define the precedence of your infix operators with `prec.left`?

**Reply by katawful** [reply]
That solved it, thanks for the help!
